### PR TITLE
Reduce the timing tests complexity

### PR DESCRIPTION
**Pull request by k-stachowiak**
## Description
This PR reduces the complexity of the timing tests to account to the degree to which we may trust underlying platforms.

Fixes https://github.com/ARMmbed/mbedtls/issues/1517

## Status
**READY**

## Requires Backporting
YES - 2.16: #2385 - 2.7: #2384

### CODE REVIEW

**Comment by k-stachowiak** [comment]
retest

**Comment by k-stachowiak** [comment]
A DTLS related test failure recorded, but timing tests pased. Will test again.

**Comment by k-stachowiak** [comment]
retest

**Comment by k-stachowiak** [comment]
retest

File: ChangeLog [hunk 1]
```diff
@@ -22,6 +22,8 @@ Changes
      Inserted as an enhancement for #1371
    * Add support for alternative CSR headers, as used by Microsoft and defined
      in RFC 7468. Found by Michael Ernst. Fixes #767.
+   * Reduced the complexity of the timing tests, as they were failing in less
+     reliable environments.
```
Comment (mpg) [inline ChangeLog]: I'd rephrase as "as they were assuming more than the underlying OS actually guarantees".

**Comment by mpg** [comment]
I'm applying "needs: backport" as it's an important fix to our CI.

**Comment by mpg** [comment]
As agreed with @sbutcher-arm I'm approving this for design.

File: ChangeLog [hunk 2]
```diff
@@ -22,6 +22,8 @@ Changes
      Inserted as an enhancement for #1371
    * Add support for alternative CSR headers, as used by Microsoft and defined
      in RFC 7468. Found by Michael Ernst. Fixes #767.
+   * Reduced the complexity of the timing tests, as they were assuming more than
```
Comment (hanno-arm) [inline ChangeLog]: Minor: Use imperative style here, i.e. `Reduce the complexity of the timing tests. They were previously assuming more than the underlying OS actually guarantees.`

File: tests/suites/test_suite_timing.function [hunk 3]
```diff
@@ -54,350 +17,52 @@ static int timers_are_badly_broken = 0;
...
-    /* A 0-second alarm should fire quickly, but we don't guarantee that it
-       fires immediately, so mbedtls_timing_alarmed may or may not be set at
-       this point. */
-
-    /* Busy-wait until the alarm rings */
-    do
-    {
-        ++iterations;
-        millis = mbedtls_timing_get_timer( &timer, 0 );
+    if( seconds == 0 ) {
```
Comment (hanno-arm) [inline tests/suites/test_suite_timing.function]: Minor: Break the line before `{`.

**Comment by k-stachowiak** [comment]
@hanno-arm Thanks for your comments.
@hanno-arm, @mpg thanks for lightning fast re-reviews. I have applied equivalent comments in the backports. Could you please review them as well?

**Comment by mpg** [comment]
@k-stachowiak @Patater should we side-port this to mbed-crypto?

**Comment by k-stachowiak** [comment]
@mpg I don't think I can speak for @Patater but it's definitely possible, and if it is desired, I may perform such a port.
